### PR TITLE
GEOPY-500: Create a geoapps "minimal install"

### PR DESCRIPTION
**Pull request by benk-mira**
**GEOPY-500 - Create a geoapps "minimal install"**
Addition of an environment-GA.yml and install_ga.bat files that define a minimal install of geoapps for inversion.  Includes restructing of utils and imports in geoapps.  See also PRs of geoana, simpeg for accompanying changes

### CODE REVIEW

**Review by sebhmg — CHANGES_REQUESTED** [review]
I am not pretending to review all the changes here ! 😉 

please see some suggestions and question.

Also `pip_requirements.txt` is empty and seems unused.

File: geoapps/utils/importing.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+
+def soft_import(package, objects=None, interrupt=False):
+
+    packagename = package.split(".")[0]
+    packagename = "gdal" if packagename == "osgeo" else packagename
+    err = (
+        f"Module '{packagename}' is missing from the environment. "
+        f"Consider installing with: 'conda install -c conda-forge {packagename}'"
+    )
```
[the remark this answers is not in the document]
Comment (benk-mira) [reply geoapps/utils/importing.py:19]: Done

File: geoapps/utils/io.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+from geoapps.shared_utils.utils import rotate_xy
+from geoapps.utils import soft_import
+
+
+def string_2_list(string):
+    """
+    Convert a list of numbers separated by comma to a list of floats
+    """
+    return [float(val) for val in string.split(",") if len(val) > 0]
```
[the remark this answers is not in the document]
Comment (benk-mira) [reply geoapps/utils/io.py:23]: done

File: geoapps/utils/general.py [hunk 3]
```diff
@@ -0,0 +1,205 @@
+#  Copyright (c) 2022 Mira Geoscience Ltd.
```
[the remark this answers is not in the document]
Comment (benk-mira) [reply geoapps/utils/general.py:1]: Done

File: geoapps/utils/io.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (benk-mira) [reply geoapps/utils/io.py:23]: actually - moved to new utils.string as per above comment

File: tests/utils_test.py [hunk 4]
```diff
@@ -452,140 +455,140 @@ def test_detrend_xy():
...
-    np.testing.assert_array_equal(tensor_2.cell_centers, tensor_1.cell_centers)
-    np.testing.assert_array_equal(tensor_2.origin, tensor_1.origin)
-
-    # Conclusion - can't expect to recover the original block_model.  There isn't
-    # a single block model that is equivalent to the tensor mesh.  The circular
-    # blockModel -> TensorMesh -> blockModel should pass for a bottom soutwest corner
-    # positive cell delimiters original block model though.
-
-
-def blockmodel_tensormesh_equivalency(ws, origin, x, y, z, ignore_origin=False):
```
[the remark this answers is not in the document]
Comment (benk-mira) [reply tests/utils_test.py:491]: yes it's gone.  I found a handful of utilities that were not being used in the code.  @domfournier will go through these during the code review and rescue anything really useful.

File: geoapps/utils/plotting.py [hunk 5]
```diff
@@ -574,3 +574,39 @@ def plot_curve(objects):
...
+    input_string: str
+        Input string value of type `val1:val2:ii` and/or a list of values `val3, val4`
+
+
+    Return
+    ------
+    list of floats
+        Corresponding list of values in float format
+
+    """
```
Comment (sebhmg) [inline geoapps/utils/plotting.py:594]: **suggestion:** use a consistent  doc style (e.g. `sorted_alphanumeric_list()`)

File: geoapps/utils/statistics.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+#  (see LICENSE file at the root of this source code package).
+
+import numpy as np
+from sklearn.neighbors import KernelDensity
+
+
+def random_sampling(
+    values, size, method="histogram", n_bins=100, bandwidth=0.2, rtol=1e-4
+):
+    """
```
Comment (sebhmg) [inline geoapps/utils/statistics.py:15]: **suggestion:** use a consistent doc style (e.g. sorted_alphanumeric_list())

**Review by sebhmg — CHANGES_REQUESTED** [review]
one last minor thing about the style for docstring

please see comments

File: geoapps/utils/plotting.py (hunk 5 above)
Comment (benk-mira) [reply geoapps/utils/plotting.py:594]: done

File: geoapps/utils/statistics.py (hunk 6 above)
Comment (benk-mira) [reply geoapps/utils/statistics.py:15]: done

File: geoapps/utils/statistics.py [hunk 7]
```diff
@@ -16,18 +16,10 @@ def random_sampling(
     Perform a random sampling of the rows of the input array based on
     the distribution of the columns values.
 
-    Parameters
-    ----------
+    :param values: Input array of values N x M, where N >> M
+    :param size: Number of indices (rows) to be extracted from the original array
```
Comment (sebhmg) [inline geoapps/utils/statistics.py:20]: preserve type in docstring, or, better, have it as type hint on the signature
```
def random_sampling(
    values: npt.NDArray[np.float64], size: int, method="histogram", n_bins=100, bandwidth=0.2, rtol=1e-4
) -> npt.NDArray[np.int_]:
```

See typing annotations for numpy: https://numpy.org/devdocs/reference/typing.html#api

File: geoapps/utils/plotting.py [hunk 8]
```diff
@@ -580,17 +581,11 @@ def input_string_2_float(input_string):
     """
     Function to input interval and value as string to a list of floats.
 
-    Parameter
-    ---------
-    input_string: str
-        Input string value of type `val1:val2:ii` and/or a list of values `val3, val4`
+    :param input_string: Input string value of type `val1:val2:ii` and/or
```
Comment (sebhmg) [inline geoapps/utils/plotting.py:584]: can have typing in signature if not in doc anymore
```
def input_string_2_float(input_string: str) -> npt.NDArray[np.int_] :
```

See typing annotations for numpy: https://numpy.org/devdocs/reference/typing.html#api

**Review by sebhmg — COMMENTED** [review]
non-blocking comments about preserving the type information:
- either in comment, as `:param param_type param_name: parameter description` (some example on [PyCharm doc](https://www.jetbrains.com/help/pycharm/using-docstrings-to-specify-types.html#:~:text=%3Aparam%20param_type%20param_name%3A))
- or using  typing in signature

**Review by sebhmg — APPROVED** [review]

**Review by sebhmg — APPROVED** [review]

**Review by sebhmg — APPROVED** [review]

File: tests/utils_test.py [hunk 9]
```diff
@@ -452,140 +455,140 @@ def test_detrend_xy():
...
-
-    ws = Workspace(os.path.join(tmp_path, "test.geoh5"))
-
-    # Generate a 3D array
-    nx, ny, nz = 6, 6, 5
-
-    x = np.linspace(0, 10, nx)
-    y = np.linspace(0, 5, ny)
-    z = np.linspace(0, 8, nz)
+# def test_blockmodel_nonuniqueness(tmp_path):
```
[the remark this answers is not in the document]
Comment (benk-mira) [reply tests/utils_test.py]: removed

**Review by sebhmg — DISMISSED** [review]

**Comment by sebhmg** [comment]
see continuation in https://github.com/MiraGeoscience/geoapps/pull/341

**Pull request closed without merging**